internal/impl/opensearch/output.go

```diff
@@ -0,0 +1,454 @@
+	return nil
+}
+
+type pendingBulkIndex struct {
+	Action   string
+	Index    string
+	Pipeline string
+	Routing  string
+	Type     string
+	Doc      interface{}
+	ID       string
+}
+
+// WriteBatch will attempt to write a message to OpenSearch, wait for
+// acknowledgement, and returns an error if applicable.
+func (e *OpenSearch) WriteBatch(ctx context.Context, msg message.Batch) error {
+	return e.Write(msg)
+}
+
+// Write will attempt to write a message to OpenSearch, wait for
+// acknowledgement, and returns an error if applicable.
+func (e *OpenSearch) Write(msg message.Batch) error {
+	if e.client == nil {
+		return component.ErrNotConnected
+	}
+
+	requests := make([]*pendingBulkIndex, msg.Len())
+	if err := msg.Iter(func(i int, part *message.Part) error {
+		jObj, ierr := part.AsStructured()
+		if ierr != nil {
+			e.log.Errorf("Failed to marshal message into JSON document: %v\n", ierr)
+			return fmt.Errorf("failed to marshal message into JSON document: %w", ierr)
+		}
+		requests[i] = &pendingBulkIndex{
+			Action:   e.actionStr.String(i, msg),
+			Index:    e.indexStr.String(i, msg),
+			Pipeline: e.pipelineStr.String(i, msg),
+			Routing:  e.routingStr.String(i, msg),
+			Type:     e.typeStr.String(i, msg),
+			Doc:      jObj,
+			ID:       e.idStr.String(i, msg),
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	start := time.Now()
+	b, _ := opensearchutil.NewBulkIndexer(opensearchutil.BulkIndexerConfig{
+		FlushInterval: time.Second * 1,
+		FlushBytes:    0,
+		Client:        e.client,
+	})
+	ctx := context.Background()
+	for _, v := range requests {
+		bulkReq, err := e.buildBulkableRequest(v)
+		if err != nil {
+			return err
+		}
+		_ = b.Add(ctx, *bulkReq)
+	}
+
+	_ = b.Close(ctx)
+
+	biStats := b.Stats()
+	dur := time.Since(start)
+
+	if biStats.NumFailed > 0 {
+		e.log.Errorf(
+			"Indexed [%s] documents with [%s] errors in %s (%s docs/sec)",
+			humanize.Comma(int64(biStats.NumFlushed)),
+			humanize.Comma(int64(biStats.NumFailed)),
+			dur.Truncate(time.Millisecond),
+			humanize.Comma(int64(1000.0/float64(dur/time.Millisecond)*float64(biStats.NumFlushed))),
+		)
+	} else {
+		e.log.Infof(
+			"Sucessfuly indexed [%s] documents in %s (%s docs/sec)",
+			humanize.Comma(int64(biStats.NumFlushed)),
+			dur.Truncate(time.Millisecond),
+			humanize.Comma(int64(1000.0/float64(dur/time.Millisecond)*float64(biStats.NumFlushed))),
+		)
+	}
+
+	return nil
+}
+
+// Close shuts down the OpenSearch writer and stops processing messages.
+func (e *OpenSearch) Close(context.Context) error {
+	return nil
+}
+
+// Build a bulkable request for a given pending bulk index item.
+func (e *OpenSearch) buildBulkableRequest(p *pendingBulkIndex) (*opensearchutil.BulkIndexerItem, error) {
+	switch p.Action {
+	case "update":
+		// opensearch util needs the body root element should be "doc"
+		doc, _ := sjson.SetOptions("", "doc", p.Doc, &sjson.Options{ReplaceInPlace: true})
```

Comment:
Should we not check the error here?

Reply:
We already had it checked before calling this method. I thought it would be redundant. Let me see if I can improve